src/dumbo/src/tcp/endpoint.rs

```diff
@@ -299,15 +304,62 @@ impl Endpoint {
     }
 }
 
+fn build_response(http_version: Version, status_code: StatusCode, body: Body) -> Response {
```

Comment:
I can not find where `parse_request_bytes()` is used outside `dumbo` crate. Can you please tell where is the usage?

There are two drawbacks that I see for keeping one single `build_response`:
1) `build_response` from `dumbo` can change/be deleted and MMDS is impacted.
2) `build_response` is a wrapper over `micro_http::Request` building. Using it inside MMDS requires to make it a public method. This would blur the dumbos API by exposing a non-`dumbo` functionality.

Although they are duplicate, I would argue that keeping them duplicate enforces better crates separation, which is preferred. Please take a look at #1819 , which is introduced by a similar scenario. There, MMDS had a dependency on micro-http inner implementation, which assured that every HTTP request is a GET. However, this implementation had changed and the previous assumption did not hold anymore, resulting in that issue.

Reply:
> I can not find where `parse_request_bytes()` is used outside `dumbo` crate. Can you please tell where is the usage?

https://github.com/moricho/firecracker/commit/68b6c0ab22cfa69b20ba39e1156d2f1c668b1a2b#diff-7d882cb8cb3286a3411207ff13d0c72bR162

and the import is here: https://github.com/moricho/firecracker/commit/68b6c0ab22cfa69b20ba39e1156d2f1c668b1a2b#diff-f7e4add39fc8dbdc081b77a61f5c586bR10 :(

> There are two drawbacks that I see for keeping one single `build_response`:
> 
> 1. `build_response` from `dumbo` can change/be deleted and MMDS is impacted.
> 2. `build_response` is a wrapper over `micro_http::Request` building. Using it inside MMDS requires to make it a public method. This would blur the dumbos API by exposing a non-`dumbo` functionality.
> 
> Although they are duplicate, I would argue that keeping them duplicate enforces better crates separation, which is preferred. Please take a look at #1819 , which is introduced by a similar scenario. There, MMDS had a dependency on micro-http inner implementation, which assured that every HTTP request is a GET. However, this implementation had changed and the previous assumption did not hold anymore, resulting in that issue.

Hmm, the situations are somehow not the same; assuring that every HTTP request is a GET sounds like something MMDS-specific, therefore the check/enforcement should be done at MMDS level, here `build_response` doesn't contain any dumbo/MMDS logic, it's more of a helper function, so that's why I suggested to leave just one. The only thing that I would expect to be different between a response that is built at MMDS level and one that is built at dumbo level is related to the allowed method(s). I agree it's weird to expose a non-dumbo function via the dumbo API, I don't like that either, we should definitely avoid it, another suggestion that I had in mind when I left the comment was to move it to `micro_http` since it's a pretty basic/generic function that can build a response based on some input.
We can also keep them duplicate to be super safe, but I would've liked to avoid that.

Also, it would be great if we could avoid using `parse_request_bytes` in MMDS tests. Btw, just curious, `MethodNotAllowed` seems already tested in `test_convert_to_response()`, why do we test it again in `test_unsupported_http_method`? Also, it looks like we are testing in MMDS the `micro_http` unsupported methods. It makes sense to test them again here as a sanity check since we must make sure that only `GET` is supported, but the problem that I see is that we don't actually check in MMDS for the `NotImplemented` methods, we kind of rely on dumbo to do the proper checks and build the right response. I would drop `test_unsupported_http_method()` until we make a clear separation here between dumbo and MMDS.

Reply:
And regarding `build_response()`, the logic from it seems to be the same in `dumbo`, `mmds` and `api_server` too (https://github.com/firecracker-microvm/firecracker/blob/master/src/api_server/src/parsed_request.rs#L94). Since building a response based on some specific inputs is rather application specific, it would make the most sense to have this in a helper crate from Firecracker, i.e. the `utils` one. But this way, `utils` will have an extra dependency on `micro-http` and we don't think it's worth it. It's fine to leave `build_response()` duplicated in `mmds` and `dumbo`.

Reply:
I fixed!